### PR TITLE
SNOW-983911 Add support for querying the vector data type

### PR DESCRIPTION
**Pull request by sfc-gh-rpanchapakesan**
### Description
Added support for parsing vectors from the serialized result set, so queries that return vectors can now be made. To do so, I also had to update `execResponseRowType` to add two new fields that are now being sent back, matching the [change](https://github.com/snowflakedb/snowflake-connector-python/pull/1806) I recently made to the Python connector.

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: driver_test.go [hunk 1]
```diff
@@ -1263,6 +1264,48 @@ func testArray(t *testing.T, json bool) {
 	})
 }
 
+func TestVector(t *testing.T) {
```
Comment (sfc-gh-pfus) [inline driver_test.go:1267]: Do we need to enable some parameter for this to work?

File: driver_test.go [hunk 2]
```diff
@@ -1263,6 +1264,48 @@ func testArray(t *testing.T, json bool) {
...
+		var gotInt []int32
+		if err := rowsInt.Scan(Vector(&gotInt)); err != nil {
+			t.Error(err)
+		}
+		wantInt := []int32{1, 2, 3}
+		if !reflect.DeepEqual(gotInt, wantInt) {
+			t.Errorf("incorrect vector deserialized: got %v, want %v", gotInt, wantInt)
+		}
+
+		rowsFloat := dbt.mustQuery(`select [1.1,2.2,3,4,5]::vector(float, 5)`)
```
Comment (sfc-gh-pfus) [inline driver_test.go:1292]: Can we split it into separate tests for int and float?

File: driver_test.go [hunk 3]
```diff
@@ -1263,6 +1264,48 @@ func testArray(t *testing.T, json bool) {
...
+func TestVector(t *testing.T) {
+	testVector(t, false)
+}
+
+func testVector(t *testing.T, json bool) {
+	runDBTest(t, func(dbt *DBTest) {
+		if json {
+			dbt.mustExec(forceJSON)
+		}
+		// dbt.mustExec(`create temporary table test_vector (i vector(int,3), f vector(float,3))`)
```
Comment (sfc-gh-pfus) [inline driver_test.go:1276]: Can we remove this query?

File: driver_test.go [hunk 4]
```diff
@@ -1263,6 +1264,48 @@ func testArray(t *testing.T, json bool) {
 	})
 }
 
+func TestVector(t *testing.T) {
+	testVector(t, false)
```
Comment (sfc-gh-pfus) [inline driver_test.go:1268]: We are lacking the test for JSON, right?

File: driver_test.go [hunk 5]
```diff
@@ -1263,6 +1264,48 @@ func testArray(t *testing.T, json bool) {
...
+}
+
+func testVector(t *testing.T, json bool) {
```
Comment (sfc-gh-pfus) [inline driver_test.go:1271]: Instead of such approach, we use `t.Run` - please see some examples in our tests.

File: converter.go [hunk 6]
```diff
@@ -91,10 +91,10 @@ func goTypeToSnowflake(v driver.Value, tsmode snowflakeType) snowflakeType {
 }
 
 // snowflakeTypeToGo translates Snowflake data type to Go data type.
-func snowflakeTypeToGo(dbtype snowflakeType, scale int64) reflect.Type {
-	switch dbtype {
+func snowflakeTypeToGo(rowType execResponseRowType) reflect.Type {
```
[the remark this answers is not in the document]
Comment (sfc-gh-rpanchapakesan) [reply converter.go:94]: I could parse the response type in the caller of this function and pass down some type of options struct with the vector metadata. However, since this function is only called in one place right now (and in the context of a request), it felt cleaner to have the function take in `execResponseRowType`.

File: converter.go [hunk 7]
```diff
@@ -108,8 +108,22 @@ func snowflakeTypeToGo(dbtype snowflakeType, scale int64) reflect.Type {
...
 	case booleanType:
 		return reflect.TypeOf(true)
+	case vectorType:
+		if len(rowType.Fields) != 1 {
+			logger.Errorf("invalid response row type for vector: %+v", rowType)
+			return reflect.TypeOf("")
+		}
+		switch getSnowflakeType(rowType.Fields[0].Type) {
+		case fixedType:
+			return reflect.TypeOf([]int32{})
```
[the remark this answers is not in the document]
Comment (sfc-gh-rpanchapakesan) [reply converter.go:118]: Yes, we only currently support 32 bit values.

File: driver_test.go (hunk 4 above)
Comment (sfc-gh-rpanchapakesan) [reply driver_test.go:1268]: Added tests for JSON to `old_driver_test.go`.

File: driver_test.go (hunk 5 above)
Comment (sfc-gh-rpanchapakesan) [reply driver_test.go:1271]: Oh ok, I was following the example of `testArray` and `testVariant` which call `runDBTest` without `t.Run`. I've converted this test into a table test that uses `t.Run`.

File: driver_test.go (hunk 1 above)
Comment (sfc-gh-rpanchapakesan) [reply driver_test.go]: I wasn't sure which environment these tests were running oi. However, it looks like the CI failures are due to the parameter, so I've added a statement to enable it for this test.

File: converter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply converter.go:95]: I think that separations of layers (API/mapping) is good :) I would prefer to have API parsing closer to API functions, and this function preserve with just types at it was before.

File: converter.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply converter.go:119]: Can we mention V1 in the commit message (so for now - in PR name only)?

File: converter.go [hunk 8]
```diff
@@ -610,6 +646,33 @@ func arrowToValue(
 			}
 		}
 		return err
+	case vectorType:
+		vectorData := srcValue.(*array.FixedSizeList)
+		datatype := vectorData.DataType().(*arrow.FixedSizeListType)
```
Comment (sfc-gh-pfus) [inline converter.go:651]: Does this work only for arrow? What should happen if a client uses `ALTER SESSION SET GO_QUERY_RESULT_FORMAT = JSON`?

File: driver_test.go [hunk 9]
```diff
@@ -1263,6 +1264,102 @@ func testArray(t *testing.T, json bool) {
 	})
 }
 
+func TestVectorInt(t *testing.T) {
+	testVectorInt(t, false)
```
Comment (sfc-gh-pfus) [inline driver_test.go:1268]: How about testing with JSON?

File: driver_test.go [hunk 10]
```diff
@@ -1263,6 +1264,102 @@ func testArray(t *testing.T, json bool) {
...
+			want:  []int32{-5},
+		},
+		{
+			msg:   "3 element vector",
+			query: `select [1,2,-3]::vector(int, 3)`,
+			want:  []int32{1, 2, -3},
+		},
+	}
+	runDBTest(t, func(dbt *DBTest) {
+		dbt.mustExec("alter session set enable_vector_data_type='Enable'")
```
Comment (sfc-gh-pfus) [inline driver_test.go:1289]: We use uppercase for parameters, can you change it?

File: driver_test.go (hunk 4 above)
Comment (sfc-gh-pfus) [reply driver_test.go]: I believe that adding it close to the Arrow tests is better for now.

**Review by sfc-gh-pfus — CHANGES_REQUESTED** [review]

File: datatype.go [hunk 11]
```diff
@@ -131,6 +135,40 @@ func dataTypeMode(v driver.Value) (tsmode snowflakeType, err error) {
 	return tsmode, nil
 }
 
+type vectorElements interface {
```
Comment (sfc-gh-pfus) [inline datatype.go:138]: I'm wondering about one thing. In the future we are expected to implement structured types, which contains arrays. Array is a very similar structure to vector, just more generic I'd say. Maybe we can prepare the code here to use these arrays instead of native backend type as vector?

File: driver_test.go [hunk 12]
```diff
@@ -1263,6 +1264,102 @@ func testArray(t *testing.T, json bool) {
...
+
+func testVectorInt(t *testing.T, json bool) {
+	tests := []struct {
+		msg   string
+		query string
+		want  []int32
+	}{
+		{
+			msg:   "1 element vector",
+			query: `select [-5]::vector(int, 1)`,
```
Comment (sfc-gh-pfus) [inline driver_test.go:1279]: Can we also have a test when the response contains empty/null value?
